### ISSUE TITLE
Static LibCurl

### ISSUE THREAD

**Issue by muv1** [ldc-developers/ldc#3376]
For some reason a static cURL library is not currently offered, only a shared
library:

~~~
lib\curl.lib
lib\libcurl.dll
~~~

So I compiled my own:

~~~
git clone --depth 1 git://github.com/curl/curl
cd curl
mingw32-make -C lib -f Makefile.m32
~~~

but it fails with `std.net.curl`:

~~~
$ cat std_net_curl.d
import std.net.curl;
void main() {
   auto s1 = "http://speedtest.lax.hivelocity.net";
   s1.get;
}

$ ldc2 std_net_curl.d wldap32.lib libcurl.lib
$ ./std_net_curl

std.net.curl.CurlException@std\net\curl.d(4223): Failed to load curl, tried
"libcurl.dll", "curl.dll".
~~~

and it fails with `etc.c.curl`:

~~~
$ cat etc_c_curl.d
import etc.c.curl;
void main() {
   curl_easy_init();
}

$ ldc2 etc_c_curl.d wldap32.lib libcurl.lib
lld-link: error: libcurl.lib(easy.o): invalid symbol index in addrsig section
~~~

what am I doing wrong here?

**Comment by kinke** [comment]
This isn't LDC-specific (so better ask upstream). AFAIK, `std.net.curl` only supports shared libraries anymore. [And if you want to build your own lib, then don't use MinGW, but the MS toolchain.]

**Comment by muv1** [comment]
@kinke can you please reopen? I am not able to successfully do this with DMD:

~~~d
import etc.c.curl, std.string;
void main() {
   auto s1 = "http://speedtest.lax.hivelocity.net";
   auto r1 = curl_easy_init();
   curl_easy_setopt(r1, CurlOption.url, s1.toStringz);
   curl_easy_perform(r1);
}
~~~

~~~
dmd -m64 test.d curl.lib advapi32.lib ucrtbase.lib wldap32.lib
~~~

so it appears that this is LDC specific.

**Comment by kinke** [comment]
You can always reopen yourself. - What has changed since https://issues.dlang.org/show_bug.cgi?id=20690 then? Please upload your curl.lib somewhere, so that I can take a look myself.

**Comment by muv1** [comment]
So it still doesnt work with `std.net.curl`, but I think thats just a problem
with the library itself. But I am able to get `etc.c.curl` working now with
DMD and LDC.

With LDC the trick was doing `ldc2 -c`, then doing `lld-link /OPT:NOICF`. I
think DMD linker adds that flag by default, maybe LDC should as well?

<https://github.com/muv1/autumn/tree/master/content/talks/d-curl-static>

**Comment by kinke** [comment]
Our `/OPT:ICF` (except with `-disable-linker-strip-dead` and/or `-g`) hasn't caused any trouble until now; maybe it depends on how the curl lib is built. You can also just override it with a `-L/OPT:NOICF` in the cmdline, or try the MS linker.

**Comment by muv1** [comment]
I am basically just doing this to build cURL:

~~~
git clone --depth 1 git://github.com/curl/curl
cd curl
mingw32-make -C lib -f Makefile.m32
~~~

**Comment by kinke** [comment]
And I'm using this (i.e., MS toolchain only) for the regular bundled DLL + import lib, in the `winbuild` dir, after building static zlib libraries:
```
# need to add `/DDONT_USE_RECV_BEFORE_SEND_WORKAROUND` to CFLAGS in MakefileBuild.vc!
nmake /f Makefile.vc mode=dll VC=15 WITH_DEVEL=C:\LDC\curl-7.65.3\deps WITH_ZLIB=static GEN_PDB=no DEBUG=no MACHINE=x64 RTLIBCFG=static
```

**Comment by muv1** [comment]
I will try the build from here:

<https://curl.haxx.se/windows>

today or tomorrow. Honestly I am pretty happy I just got it working finally. I
have been trying on and off for over 2 years to build a D program with static
cURL.

**Comment by kinke** [comment]
That's a MinGW build and comes with a libssh2 dependency etc. - I built my own static curl lib using above command (but `mode=static`), then tested with your example and `ldc2 curl.d libcurl_a.lib wldap32.lib crypt32.lib normaliz.lib`, links & works fine with both MS and LLD linkers.

**Comment by muv1** [comment]
I have been building with LLVM (Clang):

<https://github.com/mstorsjo/llvm-mingw>

I will try again with GCC.

**Comment by kinke** [comment]
I've uploaded my builds (x64 + x86, DLL + import lib + static lib): https://github.com/ldc-developers/mingw-w64-libs/releases/download/v7.0.0/libcurl-7.69.1-zlib-static-ipv6-sspi-winssl.7z

**Comment by kinke** [comment]
And when specifying the newly included .exp file when linking the executable (`-Lpath\to\curl.exp`), the executable exports the symbols, and your initial `std.net.curl` example works just fine too.

**Comment by kinke** [comment]
Should be resolved for good with #3378; please reopen in case it doesn't work for you.

**Comment by muv1** [comment]
Thank you! Sorry for not getting back with you sooner, I was trying to get a personal build that would work, but it seems only the MSVC compiler is working without tweaks. If memory serves your version did work as expected, but I am still doing testing. Thanks again!

<https://github.com/muv1/curl>

**Comment by kinke** [comment]
[I'm pretty sure it works with the clang compiler too, but you'll need to target the MSVC runtime, not the MinGW one.]

**Comment by muv1** [comment]
I havent tested with `clang-cl` yet, but I wanted you to know that I have now
tested with the LibCurl here:

<https://curl.haxx.se/windows>

just to see what would happen. With both LDC and DMD, among other errors I get
this:

~~~
lld-link: error: undefined symbol: gettimeofday
~~~

and after doing a search it seems neither project has any libraries that define
that, so it seems its not possible to use that build of LibCurl with D.

**Comment by nu8** [comment]
> 
> 
> And when additionally specifying the newly included .exp file when linking the executable (`-Lpath\to\curl.exp`), the executable exports the symbols, and your initial `std.net.curl` example works just fine too.

Can you clarify? I get this:

~~~
PS D:\Git\gulf\chapter-4> ldc2 https.d curl_a.lib -Lcurl.exp
Error: unrecognized file extension exp
~~~

**Comment by kinke** [comment]
@nu: I guess you're using LLD and that it apparently doesn't support .exp directly. The MS linker does.

**Comment by nu8** [comment]
@kinke I am not using any special software, just this:

https://github.com/ldc-developers/ldc/releases/download/v1.22.0-beta2/ldc2-1.22.0-beta2-windows-x64.7z

**Comment by kinke** [comment]
Let me rephrase: `-v` should show an lld-link cmdline, as you apparently don't have a Visual C++ installation. MS link.exe definitely works.

**Comment by nu8** [comment]
No, it doesnt:

~~~
PS C:\> ldc2 -v https.d curl_a.lib -Lcurl.exp
binary    C:\ldc2\bin\ldc2.exe
version   1.22.0-beta2 (DMD v2.092.0, LLVM 10.0.0)
config    C:\ldc2\etc\ldc2.conf (x86_64-pc-windows-msvc)
predefs   LDC all D_Version2 assert D_ModuleInfo D_Exceptions D_TypeInfo X86_64 D_InlineAsm_X86_64 D_HardFloat LittleEndian D_LP64 D_PIC Windows Win64 CRuntime_Microsoft CppRuntime_Microsoft LDC_LLVM_1000
Error: unrecognized file extension exp
~~~

**Comment by kinke** [comment]
Lol, this has to do with the powershell (the color highlighting helps). Powershell for some reason needs the -Lcurl.exp in quotes.

**Comment by nu8** [comment]
ok that worked, thank you. But how did you create `curl.exp`?

**Comment by kinke** [comment]
See https://github.com/ldc-developers/ldc/pull/3378 fow how it's been built.

**Comment by nu8** [comment]
I finally did it. Using this file:

~~~d
import std.net.curl, std.stdio;
void main() {
   auto s1 = "https://speedtest.lax.hivelocity.net";
   auto s2 = s1.get;
   s2.write;
}
~~~

I build cURL like this:

~~~ps1
git clone --depth 1 git://github.com/curl/curl
Set-Location curl/lib
mingw32-make `
-f Makefile.m32 `
-j 5 `
CFG=-winssl `
CURL_CFLAG_EXTRAS=-fno-addrsig
Copy-Item libcurl.a curl.lib
~~~

I never noticed before, but building cURL this way automatically creates
`libcurl.def`. Then build program like this:

~~~
ldc2 a.d `
crypt32.lib `
curl\lib\curl.lib `
curl\lib\libcurl.def `
wldap32.lib
~~~

Its a fully static, HTTPS capable D program, without needing any Microsoft
tools, only this:

<https://github.com/mstorsjo/llvm-mingw>

I did get a bunch of these:

~~~
lld-link: warning: duplicate /export option: curl_global_init
lld-link: warning: duplicate /export option: curl_global_init_mem
lld-link: warning: duplicate /export option: curl_global_cleanup
lld-link: warning: duplicate /export option: curl_easy_init
lld-link: warning: duplicate /export option: curl_easy_perform
~~~

but everything still seems to work ok.

**Comment by kinke** [comment]
I'm pretty sure it's not fully static, but requires an installed MSVC runtime.

**Comment by nu8** [comment]
> I'm pretty sure it's not fully static, but requires an installed MSVC runtime.

I mean yeah:

~~~
PS C:\> llvm-objdump.exe -p a.exe | ag DLL
    DLL Name: vcruntime140.dll
    DLL Name: KERNEL32.dll
    DLL Name: ADVAPI32.dll
    DLL Name: WS2_32.dll
    DLL Name: ucrtbase.dll
    DLL Name: CRYPT32.dll
    DLL Name: WLDAP32.dll
~~~

but who cares? That is a result of LDC, not of the library I built:

~~~
lld-link /NOLOGO /OPT:REF /OPT:ICF /DEFAULTLIB:vcruntime140 /OUT:a.exe a.obj
/DEF:curl\lib\libcurl.def crypt32.lib curl\lib\curl.lib wldap32.lib
C:/ldc2/bin/../lib\ldc_rt.builtins.lib /LIBPATH:C:/ldc2/bin/../lib
/LIBPATH:C:/ldc2/bin/../lib/mingw phobos2-ldc.lib druntime-ldc.lib kernel32.lib
user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib
comdlg32.lib advapi32.lib oldnames.lib legacy_stdio_definitions.lib
~~~

**Comment by kinke** [comment]
> That is a result of LDC, not of the library I built:

Yes, LDC's EH depends on MSVC functionality. Which in the context of non-MS toolchains means that none of them can really produce fully static binaries (when using EH etc.), and that's my point.

> but who cares?

Well, every machine running the binary needs a 3rd-party MS package - but great, at least no MS tools were needed to build the binary! ;) I don't understand the apparent allergic reaction of some people when it comes to installing an MSVC toolchain, which yes, takes a couple of GBs when installed and might take some minutes to install. But that's the 'pain' for one person which allows to ship truly dependency-free binaries for numerous users, and operate *with* the standard dev ecosystem on Windows.

**Comment by nu8** [comment]
Here you go:

~~~
ldc2 a.d `
crypt32.lib `
curl\lib\curl.lib `
curl\lib\libcurl.def `
ucrtbase.lib `
wldap32.lib
~~~

Result:

~~~
PS C:\> llvm-objdump.exe -p a.exe | ag DLL
    DLL Name: ucrtbase.dll
    DLL Name: KERNEL32.dll
    DLL Name: ADVAPI32.dll
    DLL Name: WS2_32.dll
    DLL Name: CRYPT32.dll
    DLL Name: WLDAP32.dll
~~~

and `ucrtbase.dll` is preinstalled with Windows 10.

**Comment by nu8** [comment]
To anyone interested, I discovered what I think is a better way to solve this
problem. Instead of trying to find a proper build or cURL, or trying to build
a D compatible cURL, another package is available that can utilize WinInet with
Windows:

https://github.com/CyberShadow/ae

This utilizes the existing Windows DLL `wininet.dll` instead of needing any cURL
library. I have some build files here:

https://github.com/nu8/gulf/tree/master/d-https

AE can do HTTPS as well.
